bravado/mapping/operation.py

```diff
@@ -1,33 +1,72 @@
-from functools import partial
 import logging
 
-from yelp_uri import urllib_utf8
-
 from bravado import swagger_type
-from bravado.mapping.docstring import create_operation_docstring, \
-    docstring_property
-from bravado.mapping.param import validate_and_add_params_to_request
 from bravado.response import post_receive, HTTPFuture
+from bravado.exception import SwaggerError
+from bravado.mapping.param import Param, marshal_param
 
 log = logging.getLogger(__name__)
 
 
-# TOOD: swagger_type.get_swagger_type() is called so many times (across this
-#       module and param.py. Refactor state in Param class
-
 class Operation(object):
     """Perform a request by taking the kwargs passed to the call and
     constructing an HTTP request.
+
+    :type swagger_spec: :class:`Spec`
+    :param path_name: path of the operation. e.g. /pet/{petId}
+    :param http_method: get/put/post/delete/etc
+    :param operation_spec: operation specification in dict form
     """
-    def __init__(self, spec, path_name, http_method, operation_dict):
-        self.spec = spec
+    def __init__(self, swagger_spec, path_name, http_method, operation_spec):
+        self.swagger_spec = swagger_spec
         self.path_name = path_name
         self.http_method = http_method
-        self.operation_dict = operation_dict
-        self._operation_id = None  # use @property getter
-        # lazy docstring
-        self.__doc__ = docstring_property(
-            partial(create_operation_docstring, self))
+        self.operation_spec = operation_spec
+
+        # generated by @property when necessary since this is optional
+        self._operation_id = None
+
+        # (key, value) = (param name, Param)
+        self.params = {}
+
+    @classmethod
+    def from_spec(cls, swagger_spec, path_name, http_method, operation_spec):
+        """
+        Creates a :class:`Operation` and builds up its list of :class:`Param`s
+
+        :param swagger_spec: :class:`Spec`
+        :param path_name: path of the operation. e.g. /pet/{petId}
+        :param http_method: get/put/post/delete/etc
+        :param operation_spec: operation specification in dict form
+        :rtype: :class:`Operation`
+        """
+        op = cls(swagger_spec, path_name, http_method, operation_spec)
+        op.build_params()
+        return op
+
+    @property
+    def resolved_param_specs(self):
+        """
+        Returns param_specs for the current operation with immediate $refs
+        resolved.
+
+        :return: list of dicts representing parameter specs
+        """
+        specs = []
+        for param_spec in self.operation_spec.get('parameters', []):
+            resolved_param_spec = self.swagger_spec.resolve(param_spec)
+            specs.append(resolved_param_spec)
+        return specs
+
+    def build_params(self):
```

Comment:
-1 on mutating this class after it's constructed.

Can this be done as a module level function and just pass in `params` to the constructor?  I guess this applies to `resolved_param_specs` as well. Neither of these are referenced on the instance itself after it's created.

That way the logic for building an `Operation` is clearly separated from the functionality of the operation.